src/main/java/org/disq_bio/disq/HtsjdkReadsRddStorage.java

```diff
@@ -0,0 +1,280 @@
+   *     the header
+   */
+  public HtsjdkReadsRdd read(String path) throws IOException {
+    return read(path, null);
+  }
+
+  /**
+   * Read reads from the given path, using the given traversal parameters to filter the reads. The
+   * input files may be in any format (BAM/CRAM/SAM).
+   *
+   * @param path the file or directory to read from
+   * @param traversalParameters parameters that determine which reads should be returned, allows
+   *     filtering by interval
+   * @param <T> the type of Locatable for specifying intervals
+   * @return a {@link HtsjdkReadsRdd} that allows access to the reads
+   * @throws IOException if an IO error occurs while determining the format of the files
+   */
+  public <T extends Locatable> HtsjdkReadsRdd read(
+      String path, HtsjdkReadsTraversalParameters<T> traversalParameters) throws IOException {
+
+    FileSystemWrapper fileSystemWrapper =
+        useNio ? new NioFileSystemWrapper() : new HadoopFileSystemWrapper();
+
+    String firstSamPath;
+    if (fileSystemWrapper.isDirectory(sparkContext.hadoopConfiguration(), path)) {
+      firstSamPath =
+          fileSystemWrapper.firstFileInDirectory(sparkContext.hadoopConfiguration(), path);
+    } else {
+      firstSamPath = path;
+    }
+    SamFormat samFormat = SamFormat.fromPath(firstSamPath);
+
+    if (samFormat == null) {
+      throw new IllegalArgumentException("Cannot find format extension for " + path);
+    }
+
+    AbstractSamSource abstractSamSource;
+    switch (samFormat) {
+      case BAM:
+        abstractSamSource = new BamSource(useNio);
+        break;
+      case CRAM:
+        abstractSamSource = new CramSource(useNio);
+        break;
+      case SAM:
+        abstractSamSource = new SamSource();
+        break;
+      default:
+        throw new IllegalArgumentException("File does not end in BAM, CRAM, or SAM extension.");
+    }
+
+    SAMFileHeader header =
+        abstractSamSource.getFileHeader(
+            sparkContext, path, validationStringency, referenceSourcePath);
+    JavaRDD<SAMRecord> reads =
+        abstractSamSource.getReads(
+            sparkContext,
+            path,
+            splitSize,
+            traversalParameters,
+            validationStringency,
+            referenceSourcePath);
+    return new HtsjdkReadsRdd(header, reads);
+  }
+
+  /**
+   * Write reads to a file or files specified by the given path. Write options may be specified to
+   * control the format to write in (BAM/CRAM/SAM, if not clear from the path extension), and the
+   * number of files to write (single vs. multiple).
+   *
+   * @param htsjdkReadsRdd a {@link HtsjdkReadsRdd} containing the header and the reads
+   * @param path the file or directory to write to
+   * @param writeOptions options to control aspects of how to write the reads (e.g. {@link
+   *     FormatWriteOption} and {@link FileCardinalityWriteOption}
```

Comment:
The write options are nice, but extending this could get a bit complicated - although I don't have a better solution in mind...